raft/raft.go

```diff
@@ -481,8 +484,12 @@ func (r *raft) maybeSendAppend(to uint64, sendIfEmpty bool) bool {
 			// optimistically increase the next when in StateReplicate
 			case tracker.StateReplicate:
 				last := m.Entries[n-1].Index
+				var bytes uint64
```

Comment:
Just saw this:

https://github.com/etcd-io/etcd/blob/419ee8a9c6799084a78b6188f40fd120a4d316a5/raft/raft.go#L1302-L1305

So on each heartbeat response (200ms), we'll be throwing out the oldest inflight. We kind of have to do that (or we'll be stuck if all inflight messages in fact did get dropped), but it seems like it might circumvent the protection.

Ideally, when inflights is full, the leader would send an empty append instead of freeing up a slot in inflights. An empty append is ~zero sized, so we can afford to ignore it for the purpose of inflight tracking.

To implement this, presumaby here

https://github.com/etcd-io/etcd/blob/419ee8a9c6799084a78b6188f40fd120a4d316a5/raft/raft.go#L437-L439

we would instead send an empty append instead of returning with `false`.

Btw, tried to see by which mechanism the leader would try to catch up a node that's behind. Let's say n1 is the leader, and n2 is fully caught up, but all inflight messages to n3 were dropped. There is no more activity (just regular heartbeats). What prompts the leader to re-send to n3? I thought it would be on heartbeats or heartbeat responses, but I don't see it.

Reply:
Sounds doable, will fix.

On re-sends, would the Transport implementation do any retries?

Reply:
So these lines will send entries (on heartbeat responses):
https://github.com/etcd-io/etcd/blob/419ee8a9c6799084a78b6188f40fd120a4d316a5/raft/raft.go#L1306-L1308

But they "optimistically" update the progress `Next`: https://github.com/etcd-io/etcd/blob/419ee8a9c6799084a78b6188f40fd120a4d316a5/raft/raft.go#L484

But it's unclear whether `Next` ever comes back if the node continues to hold leadership.

Maybe the assumption is that there is something like TCP underneath, so if we are able to heartbeat then other messages including appends will slip through the stream.

Reply:
> There is no more activity (just regular heartbeats). What prompts the leader to re-send to n3? I thought it would be on heartbeats or heartbeat responses, but I don't see it.

My hypothesis after more reading is:
- We keep sending `MsgApp` indefinitely.
- Eventually we receive `MsgAppResp` with a reject.
- This [throws](https://github.com/etcd-io/etcd/blob/419ee8a9c6799084a78b6188f40fd120a4d316a5/raft/raft.go#L1242) the progress back, and results in retries.

Reply:
Fix in #14633, I think the described issue is pre-existing.